src/buffer/out/textBuffer.cpp

```diff
@@ -166,6 +166,29 @@ ROW& TextBuffer::_getRowByOffsetDirect(size_t offset)
     return *reinterpret_cast<ROW*>(row);
 }
 
+// This function is "direct" because it trusts the caller to know that this row
+// could be the scratchpad row.
+size_t TextBuffer::_estimateOffsetOfLastCommittedRowDirect() const noexcept
+{
+    if (_commitWatermark == _buffer.get())
+    {
+        // Estimation failed. Act as though we have at least one non-scratchpad row committed.
+        // The caller will be passing this offset back to GetRowByOffset or similar later, which
+        // will commit it.
+        return 1;
+    }
+    const auto lastRowOffset = ((_commitWatermark - 1) - _buffer.get()) / _bufferRowStride;
+    return gsl::narrow_cast<size_t>(lastRowOffset);

```

Comment:
nit: Purely theoretically speaking, if `_commitReadAheadRowCount` was 0, this could return 0 instead of 1. It's not really an issue at all, but this code only works because other code works in particular way. (I mean I also write tons of code like that, but... it's objectively not quite perfect IMO.)

From my personal perspective I also wouldn't consider the scratchpad row a "real" row and so I personally don't think `_estimateOffsetOfLastCommittedRowDirect` is worth keeping. `_getRowByOffsetDirect` only exists because I wanted to deduplicate the two row accessors, but I've also considered comitting the first viewport worth of rows eagerly so that we can just blindly assume that `_buffer.get()` is always committed and thus return that pointer as the scratchpad row instantly.

In short, I'd inline `_estimateOffsetOfLastCommittedRowDirect` into `_estimateOffsetOfLastCommittedRow` and return 0 above.

Reply:
SGTM, thanks!